pkg/generate/generator/strategyref.go

```diff
@@ -0,0 +1,167 @@
+	return &BuildStrategyRefGenerator{
+		gitRepository:     git.NewRepository(),
+		dockerfileFinder:  dockerfile.NewFinder(),
+		dockerfileParser:  dockerfile.NewParser(),
+		sourceDetectors:   sourceDetectors,
+		imageRefGenerator: NewImageRefGenerator(),
+	}
+}
+
+// FromSourceRef creates a build strategy from a source reference
+func (g *BuildStrategyRefGenerator) FromSourceRef(srcRef app.SourceRef) (*app.BuildStrategyRef, error) {
+
+	// Download source locally first if not available
+	if len(srcRef.Dir) == 0 {
+		if err := g.getSource(&srcRef); err != nil {
+			return nil, err
+		}
+	}
+
+	// Detect a Dockerfile
+	context, found, err := g.detectDockerFile(srcRef.Dir)
+	if err != nil {
+		return nil, err
+	}
+	if found {
+		return g.FromSourceRefAndDockerContext(srcRef, context)
+	}
+
+	// Detect a STI repository
+	sourceInfo, ok := g.sourceDetectors.DetectSource(srcRef.Dir)
+	if !ok {
+		return nil, errors.CouldNotDetect
+	}
+	builderImage, err := g.imageForSourceInfo(sourceInfo)
+	if err != nil {
+		return nil, err
+	}
+	return g.FromSTIBuilderImage(builderImage)
+}
+
+// FromSourceRefAndDockerContext generates a BuildStrategyRef from a source ref and context path
+func (g *BuildStrategyRefGenerator) FromSourceRefAndDockerContext(srcRef app.SourceRef, context string) (*app.BuildStrategyRef, error) {
+	// Download source locally first if not available
+	if len(srcRef.Dir) == 0 {
+		if err := g.getSource(&srcRef); err != nil {
+			return nil, err
+		}
+	}
+
+	// Look for Dockerfile in repository
+	file, err := os.Open(filepath.Join(srcRef.Dir, context, "Dockerfile"))
+	if err != nil {
+		return nil, err
+	}
+
+	dockerFile, err := g.dockerfileParser.Parse(file)
+	if err != nil {
+		return nil, err
+	}
+
+	parentImageName, ok := dockerFile.GetDirective("FROM")
+	if !ok {
+		return nil, errors.InvalidDockerfile
+	}
+
+	return g.FromDockerContextAndParent(context, parentImageName[0])
+
+}
+
+// FromContextAndParent generates a build strategy ref from a context path and parent image name
+func (g *BuildStrategyRefGenerator) FromDockerContextAndParent(context string, parentImageName string) (*app.BuildStrategyRef, error) {
+	var parentImage, err = g.imageRefGenerator.FromName(parentImageName)
+	if err != nil {
+		return nil, err
+	}
+	return &app.BuildStrategyRef{
+		IsDockerBuild: true,
+		Base:          parentImage,
+		DockerContext: context,
+	}, nil
+}
+
+// FromSTIBuilderImage generates a build strategy from a builder image ref
+func (g *BuildStrategyRefGenerator) FromSTIBuilderImage(image *app.ImageRef) (*app.BuildStrategyRef, error) {
+	return &app.BuildStrategyRef{
+		IsDockerBuild: false,
+		Base:          image,
+	}, nil
+}
+
+func (g *BuildStrategyRefGenerator) imageForSourceInfo(s *source.Info) (*app.ImageRef, error) {
+	var imageName string
+	// TODO: More sophisticated matching
+	switch s.Platform {
+	case "Ruby":
+		imageName = "openshift/ruby-20-centos"
```

Comment:
Also I think this is something that 'Detector' should return, so the detectors could be more self-contained.